### DISCUSSION TITLE
I have an idea question about a firmware application.

### DISCUSSION THREAD

**Discussion by dixxLee** [bitcraze/discussions#511] [Q&A]
Hello.
I'm a student in South Korea.
I'm studying to throw a Crazyflie and do a Hover.
I've tried many things, but I'm still having trouble.
I am trying to inject the code to detect the launch into Crazyflie Firmware, but it is detected and the motor is armed, but it continues to fall because it is impossible to control the attitude.
Currently I injected Detect Code from power_distribution_quadcopter.c.
But this doesn't work normally.
Thinking about the reason, looking at Crazyflie's stabilizer overview, even if the controller is given an althold command, the controller is running, so when the launch is detected, the setpoint is incorrectly entered, so it is considered impossible to control the attitude.
So, at the moment when the motor becomes an arm after detection, I want to inject set points of Roll, Pitch, and Yaw to 0, and use z-ranger to control altitude.
Where should I inject the Launch Detect Code so that the setpoint is injected at the time of detection?

**Comment by hmllr** [comment]
Hi :)
Unfortunately, I don't fully understand your question - what do you mean with launch? Taking off? How do you send the command to take off then? In general, to write an own application that generates setpoints I would recommend to create your own task with the app layer (see https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/). I guess especially the "push demo" can help you: https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/demos/app_push_demo
Hope this helps,
Hanna

**Reply by dixxLee** [reply]
Yes. Launch mean Taking off. It's hand launching.
When I throw Crazyflie while pressing assist mode (Pos Hold) with the remote controller, Accel data is used to detect the throw and activate the motor.

**Reply by hmllr** [reply]
okay, then I think it is best to detect this in your own task and then use the high level commander (documented here: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/commanders_setpoints/#high-level-commander) to keep the position.

**Reply by krichardsson** [reply]
You can use a "free fall" detector to figure out if it is time to start flying. 
We used to have a free fall detection earlier in the firmware but it has been removed, you can find the code [here](https://github.com/bitcraze/crazyflie-firmware/blob/a6d6f6a50f63ee4b01635f87dcde07ad0c7c0933/src/modules/src/sitaw.c) if you want to use it for inspiration.

**Reply by dixxLee** [reply]
I used the 'crtpCommanderHighLevelGetSetpoint' code in crtp_commander_high_level.c to run this code when Crazyflie is detected being  "free fall". Is this the right way?

**Reply by hmllr** [reply]
It could be - but then you need to have a high level trajectory planned. I think the easier way would be to use crtpCommanderHighLevelGoTo() to command the x,y,yaw position you already are at and the desired height.

**Reply by dixxLee** [reply]
GoTo thinks it is different from my purpose. This is because I can do hover under any conditions when I randomly. I'm still trying, but when all the accel values of X, Y, and Z have become zero, I have a motor to be Arming. And at this point the controller must run.

**Reply by hmllr** [reply]
you could also try to set a velocity setpoint directly, as in the push demo (https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/demos/app_push_demo/src/push.c#L47)?

**Reply by dixxLee** [reply]
Is this .c file that automatically works?

**Reply by hmllr** [reply]
it works automatically for the push demo it is intended for - I just referenced it as an example on how to set a hover setpoint. For stabilizing after throwing the crazyflie there is no code currently integrated, so you'll have to write it yourself - the code @krichardsson sent you can help to detect the free fall and the demo I sent can help you to set a hoversetpoint.
